src/QoreJniClassMap.cpp

```diff
@@ -883,29 +885,37 @@ void QoreJniClassMap::doFields(QoreBuiltinClass& qc, jni::Class* jc) {
 }
 
 JniExternalProgramData::JniExternalProgramData(QoreNamespace* n_jni) : jni(n_jni) {
-   assert(jni);
-   Env env(false);
+    assert(jni);
+    Env env(false);
 
-   // set up QoreURLClassLoader constructor args
-   jvalue jarg;
-   jarg.j = (long)getProgram();
-   assert(jarg.j);
+    // set up QoreURLClassLoader constructor args
+    jvalue jarg;
+    jarg.j = (long)getProgram();
+    assert(jarg.j);
 
-   // create our custom classloader
-   classLoader = env.newObject(Globals::classQoreURLClassLoader, Globals::ctorQoreURLClassLoader, &jarg).makeGlobal();
+    // create our custom classloader
+    classLoader = env.newObject(Globals::classQoreURLClassLoader, Globals::ctorQoreURLClassLoader, &jarg).makeGlobal();
 
-   // setup classpath
-   TempString classpath(SystemEnvironment::get("QORE_JNI_CLASSPATH"));
-   if (classpath)
-      addClasspath(classpath->c_str());
+    // setup classpath
+    TempString classpath(SystemEnvironment::get("QORE_JNI_CLASSPATH"));
+    if (classpath) {
+        addClasspath(classpath->c_str());
+    }
 }
 
-JniExternalProgramData::JniExternalProgramData(const JniExternalProgramData& parent) {
-   // reuse the same classLoader as the parent
-   classLoader = GlobalReference<jobject>::fromLocal(parent.classLoader.toLocal());
-
-   // copy the parent's class map to this one
-   jcmap = parent.jcmap;
+JniExternalProgramData::JniExternalProgramData(const JniExternalProgramData& parent, QoreProgram* pgm) :
+    // reuse the same classLoader as the parent
+    classLoader(GlobalReference<jobject>::fromLocal(parent.classLoader.toLocal())) {
+    // copy the parent's class map to this one
+    jcmap = parent.jcmap;
+    // find Jni namespace in new Program if present
+    const QoreNamespace* jnins = pgm->findNamespace("Jni");
+    if (jnins) {
+        jni = const_cast<QoreNamespace*>(jnins);
```

Comment:
I really don't like the `const_cast`, there is (or should be) some reason why the namespace object is returned as `const` and we should respect it (also compiler can optimize const object in some way, like copying them instead of returning reference to them etc, because they are suppose to be `const`, I'm not saying this is the case, but anyway)

what is the reason of returning the QoreNamespace as const from findNamespace method? Can we create an API where it will be returned as non-const?

Reply:
> what is the reason of returning the QoreNamespace as const from findNamespace method? Can we create an API where it will be returned as non-const?

The reason is because the namespace should generally not be changed; it is owned by the `QoreProgram` object and should generally not be modified, however this rule can be broken if the Program's parse lock is acquired before modifying the namespace, which is what the jni module does here:
```
         CurrentProgramRuntimeExternalParseContextHelper pch;
```

So we could update `QoreProgram::findNamespace()` to return a non-`const` namespace ptr, if you believe this is the better solution, because the namespace actually can be modified if the lock is also grabbed.  I'm not sure what the best solution is.  I'm also not a fan of `const_cast<>`, I feel like it's cheating somehow.

Reply:
this is now fixed in a pending commit - waiting on https://github.com/qorelanguage/qore/pull/3059 to be merged - it turns out there were no locking issues involved after all contrary to our discussion yesterday

Reply:
ok, merged